RetailCoder.VBE/Refactorings/Rename/RenameHandlerBase.cs

```diff
@@ -0,0 +1,106 @@
+    {
+        private readonly RenameModel _model;
+        private readonly IMessageBox _messageBox;
+        private List<QualifiedModuleName> _updatedQMNs;
+
+        public RenameHandlerBase(RenameModel model, IMessageBox messageBox)
+        {
+            _model = model;
+            _messageBox = messageBox;
+            _updatedQMNs = new List<QualifiedModuleName>();
+        }
+
+        public RenameModel Model { get { return _model; } }
+
+        public IMessageBox MessageBox { get { return _messageBox; } }
+
+        public void Rewrite()
+        {
+            _updatedQMNs.Distinct().ToList()
+                .ForEach(qmn => Model.State.GetRewriter(qmn).Rewrite());
+        }
+
+        public virtual void Rename() { }
+        public virtual string ErrorMessage { get; }
+
+        public void RenameUsages(Declaration target)
+        {
+            RenameUsages(target, _model.NewName);
+        }
+
+        public void RenameUsages(Declaration target, string newName)
+        {
+            var qualifiedModuleNames = new List<QualifiedModuleName>();
+            var modules = target.References.GroupBy(r => r.QualifiedModuleName);
+            foreach (var grouping in modules)
+            {
+                qualifiedModuleNames.Add(grouping.Key);
+                var rewriter = _model.State.GetRewriter(grouping.Key);
+                var module = grouping.Key.Component.CodeModule;
+                foreach (var line in grouping.GroupBy(reference => reference.Selection.StartLine))
+                {
+                    var lastSelection = Selection.Empty;
+                    foreach (var reference in line.OrderByDescending(l => l.Selection.StartColumn))
+                    {
+                        if (reference.Selection == lastSelection)
+                        {
+                            continue;
+                        }
+                        var newContent = reference.Context.GetText().Replace(reference.IdentifierName, newName);
+                        rewriter.Replace(reference.Context, newContent);
+                        lastSelection = reference.Selection;
+                    }
+                }
+            }
+            _updatedQMNs.AddRange(qualifiedModuleNames);
+        }
+
+        public void RenameDeclaration(Declaration target)
+        {
+            RenameDeclaration(target, _model.NewName);
+        }
+
+        public void RenameDeclaration(Declaration target, string newName)
+        {
+            var qualifiedModuleName = target.QualifiedName.QualifiedModuleName;
+            var component = qualifiedModuleName.Component;
+            var rewriter = _model.State.GetRewriter(target);
+            var module = component.CodeModule;
+            if (!target.DeclarationType.HasFlag(DeclarationType.Property))
+            {
+                var newContent = target.Context.GetText().Replace(target.IdentifierName, newName);
+                rewriter.Replace(target.Context, newContent);
```

Comment:
Say you have a [SubStmtContext](https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.Parsing/Grammar/VBAParser.g4#L481-L485); its `subroutineName()` method gives you a context that only includes the only token you're going to be replacing - looking at the `Rubberduck.Parsing.Symbols.Identifier` static class, there's a bunch of `string GetName` overloads, many with an `out Interval tokenInterval` parameter, which encapsulates exactly what you want to give the `TokenStreamRewriter`  =)

I think you could add overloads like:

````csharp
public static string GetName(VBAParser.SubStmtContext context, out Interval tokenInterval)
{
    return GetName(context.subroutineName(), out tokenInterval);
}

public static string GetName(VBAParser.FunctionStmtContext context, out Interval tokenInterval)
{
    return GetName(context.functionName(), out tokenInterval);
}

public static string GetName(VBAParser.PropertyGetStmtContext context, out Interval tokenInterval)
{
    return GetName(context.functionName(), out tokenInterval);
}

//...
````

You wouldn't really care for the `string`, but the `Interval` is exactly what you're after.